ugvc/mrd/srsnv_plotting_utils.py

```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+            fig_cbar.set_ticks(np.arange(0, 1, 1 / n_cat) + 1 / (2 * n_cat))
+            fig_cbar.set_ticklabels(cat_values, fontsize=12, rotation=rotation * 25, ha=ha)
+            fig_cbar.outline.set_visible(False)
+
+            # Update the start position for the next colorbar
+            start_y -= vertical_padding + rotation / (
+                rotated_padding_factor
+            )  # Adjust this value to control vertical spacing between colorbars
+
+        # Add an extra colorbar for numerical features below the others
+        cbar_num_ax = fig.add_axes([1.1, start_y - 0.03, total_width, 0.03])
+        fig_num_cbar = fig.colorbar(
+            cm.ScalarMappable(norm=None, cmap=plt.get_cmap(cmap)), cax=cbar_num_ax, orientation="horizontal"
+        )
+        fig_num_cbar.set_label("Numerical features", fontsize=12)
+        fig_num_cbar.set_ticks([0, 1])
+        fig_num_cbar.set_ticklabels(["Low value", "High value"], fontsize=12)
+        fig_num_cbar.outline.set_visible(False)
+
+    @exception_handler
+    def calc_and_plot_shap_values(
+        self,
+        output_filename_importance: str = None,
+        output_filename_beeswarm: str = None,
+        n_sample: int = 10_000,
+        plot_feature_importance: bool = True,
+        plot_beeswarm: bool = True,
+        feature_importance_kws: dict = None,
+        beeswarm_kws: dict = None,
+    ):
+        """Calculate and plot SHAP values for the model."""
+        feature_importance_kws = feature_importance_kws or {}
+        beeswarm_kws = beeswarm_kws or {}
+        # Define model, data
+        k = 0  # fold_id
+        model = self.models[k]
+        X_val = self.data_df[self.data_df["fold_id"] == k]
+        # Get SHAP values
+        logger.info("Calculating SHAP values")
+        shap_values, X_val, y_val = self._shap_on_sample(  # pylint: disable=unused-variable
+            model, X_val, n_sample=n_sample
+        )
+        logger.info("Done calculating SHAP values")
+        # SHAP feature importance
+        mean_abs_SHAP_scores = pd.Series(
+            np.abs(shap_values[:, 1, :-1] - shap_values[:, 0, :-1]).mean(axis=0), index=X_val.columns
+        ).sort_values(ascending=False)
+        mean_abs_SHAP_scores.to_hdf(self.output_h5_filename, key="mean_abs_SHAP_scores", mode="a")
+        # Plot shap scores
+        if plot_feature_importance:
+            self.plot_SHAP_feature_importance(
+                shap_values, X_val, output_filename=output_filename_importance, **feature_importance_kws
+            )
+        if plot_beeswarm:
+            self.plot_SHAP_beeswarm(shap_values, X_val, output_filename=output_filename_beeswarm, **beeswarm_kws)
+
+    def _get_trinuc_stats(self, q1: float = 0.1, q2: float = 0.9):
+        data_df = self.data_df.copy()
+        data_df["is_cycle_skip"] = data_df["is_cycle_skip"].astype(int)
+        trinuc_stats = data_df.groupby(["trinuc_context_with_alt", "label", "is_forward", "is_mixed"]).agg(
```

Comment:
These should all be constants too:
"is_cycle_skip", "trinuc_context_with_alt", "label", "is_forward", "is_mixed"

Reply:
Many or all of them already are, just need to import them